File: macos/Sources/Pomodoro/Core/Reports.swift
import Foundation

struct DayStats {
    var key = ""
    var sessions = 0
    var focusSeconds = 0
    var breakSeconds = 0
    var breaks = 0
    var completedBreaks = 0
    var shortBreaks = 0
    var longBreaks = 0
    var interruptedFocus = 0
    var interruptedBreaks = 0
    var phases = 0

    var totalSeconds: Int { focusSeconds + breakSeconds }
    var focusText: String { Fmt.reportDuration(focusSeconds) }
    var breakText: String { Fmt.reportDuration(breakSeconds) }
    var totalText: String { Fmt.reportDuration(totalSeconds) }
}

struct WeekDay: Identifiable {
    var key: String
    var label: String
    var dayNumber: Int
    var isToday: Bool
    var sessions: Int
    var breaks: Int
    var focusSeconds: Int
    var breakSeconds: Int
    var id: String { key }

    var totalSeconds: Int { focusSeconds + breakSeconds }
    var focusText: String { Fmt.reportDuration(focusSeconds) }
    var breakText: String { Fmt.reportDuration(breakSeconds) }
}

struct WeekReport {
    var startKey = ""
    var endKey = ""
    var startLabel = ""
    var endLabel = ""
    var days: [WeekDay] = []
    var sessions = 0
    var focusSeconds = 0
    var breakSeconds = 0
    var maxTotalSeconds = 0

    var focusText: String { Fmt.reportDuration(focusSeconds) }
    var breakText: String { Fmt.reportDuration(breakSeconds) }
    var averageDayText: String { Fmt.reportDuration(Double(focusSeconds) / 7) }
}

struct MonthCell: Identifiable {
    var id: Int
    var inMonth: Bool
    var key: String = ""
    var dayNumber: Int = 0
    var isToday: Bool = false
    var focusSeconds: Int = 0
    var breakSeconds: Int = 0
    var sessions: Int = 0

    var focusText: String { Fmt.reportDuration(focusSeconds) }
}

struct MonthReport {
    var key = ""
    var label = ""
    var shortLabel = ""
    var cells: [MonthCell] = []
    var sessions = 0
    var focusSeconds = 0
    var breakSeconds = 0
    var activeDays = 0
    var maxDaySeconds = 0

    var focusText: String { Fmt.reportDuration(focusSeconds) }
    var breakText: String { Fmt.reportDuration(breakSeconds) }
}

struct MonthBar: Identifiable {
    var key: String
    var label: String
    var focusSeconds: Int
    var breakSeconds: Int
    var sessions: Int
    var id: String { key }
    var totalSeconds: Int { focusSeconds + breakSeconds }
    var focusText: String { Fmt.reportDuration(focusSeconds) }
    var breakText: String { Fmt.reportDuration(breakSeconds) }
}

struct AllTimeReport {
    var sessions = 0
    var focusSeconds = 0
    var breakSeconds = 0
    var breaks = 0
    var activeDays = 0
    var currentStreak = 0
    var longestStreak = 0
    var interrupted = 0
    var firstEndedAt: EpochMillis = 0
    var lastEndedAt: EpochMillis = 0
    var months: [MonthBar] = []
    var maxChartSeconds = 0

    var focusText: String { Fmt.reportDuration(focusSeconds) }
    var breakText: String { Fmt.reportDuration(breakSeconds) }
    var averageSessionText: String {
        sessions > 0 ? Fmt.reportDuration(Double(focusSeconds) / Double(sessions)) : "0m"
    }
}

/// Report aggregation, ported from `Service.qml`. Every figure here is derived
/// from `sessions`, so the three front ends report identical numbers from the
/// same history file.
extension PomodoroService {

    func statsForDay(_ key: String) -> DayStats {
        var stats = DayStats(key: key)
        for entry in sessions where Fmt.dateKey(entry.endedAt) == key {
            let active = entry.activeSeconds
            stats.phases += 1
            if entry.phase == .focus {
                stats.focusSeconds += active
                if entry.isCompletedFocus {
                    stats.sessions += 1
                } else if active > 0 {
                    stats.interruptedFocus += 1
                }
            } else {
                stats.breakSeconds += active
                if active > 0 { stats.breaks += 1 }
                if entry.status == .completed { stats.completedBreaks += 1 }
                if entry.phase == .short && active > 0 { stats.shortBreaks += 1 }
                if entry.phase == .long && active > 0 { stats.longBreaks += 1 }
                if entry.status != .completed && active > 0 { stats.interruptedBreaks += 1 }
            }
        }
        return stats
    }

    private func clippedSegments(_ segments: [Segment],
                                 from rangeStart: EpochMillis,
                                 to rangeEnd: EpochMillis) -> [Segment] {
        segments.compactMap { segment in
            let start = max(rangeStart, segment.startedAt)
            let end = min(rangeEnd, segment.endedAt)
            guard end > start else { return nil }
            return Segment(startedAt: start, endedAt: end)
        }
    }

    /// Every phase overlapping the day, including the one in progress, sorted
    /// by start. This is what the timeline and the session list read.
    func entriesForDay(_ key: String) -> [SessionEntry] {
        guard let dayStart = Fmt.dayStartMillis(forKey: key) else { return [] }
        let dayEnd = dayStart + 86_400_000
        var rows: [SessionEntry] = []

        for entry in sessions {
            guard entry.endedAt >= dayStart, entry.startedAt < dayEnd else { continue }
            var row = entry
            row.segments = clippedSegments(entry.segments, from: dayStart, to: dayEnd)
            rows.append(row)
        }

        if phaseStartedAt > 0 {
            let now = nowMillis()
            let activeEnd = max(now, phaseStartedAt)
            if activeEnd >= dayStart && phaseStartedAt < dayEnd {
                rows.append(SessionEntry(
                    id: "active-phase",
                    phase: phase,
                    status: running ? .running : .paused,
                    startedAt: phaseStartedAt,
                    endedAt: now,
                    plannedSeconds: phasePlannedSeconds > 0 ? phasePlannedSeconds : duration(for: phase),
                    activeSeconds: phaseElapsed(at: now),
                    segments: clippedSegments(phaseSegments(at: now), from: dayStart, to: dayEnd),
                    note: phase == .focus ? activeNote : "",
                    isLive: true
                ))
            }
        }

        rows.sort { $0.startedAt < $1.startedAt }
        return rows
    }

    /// Completed focus sessions for the day, newest first.
    func sessionsForDay(_ key: String, limit: Int? = nil) -> [SessionEntry] {
        var result: [SessionEntry] = []
        for entry in entriesForDay(key).reversed() where entry.isCompletedFocus {
            result.append(entry)
            if let limit, result.count >= limit { break }
        }
        return result
    }

    func weeklyStats(anchor: Date) -> WeekReport {
        let cal = Fmt.calendar
        let today = cal.startOfDay(for: anchor)
        let monday = cal.date(byAdding: .day, value: -Fmt.mondayFirstIndex(today), to: today) ?? today
        let names = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]

        var report = WeekReport()
        for index in 0..<7 {
            guard let date = cal.date(byAdding: .day, value: index, to: monday) else { continue }
            let stats = statsForDay(Fmt.dateKey(date))
            report.days.append(WeekDay(
                key: stats.key,
                label: names[index],
                dayNumber: cal.component(.day, from: date),
                isToday: stats.key == todayKey,
                sessions: stats.sessions,
                breaks: stats.breaks,
                focusSeconds: stats.focusSeconds,
                breakSeconds: stats.breakSeconds
            ))
            report.sessions += stats.sessions
            report.focusSeconds += stats.focusSeconds
            report.breakSeconds += stats.breakSeconds
            report.maxTotalSeconds = max(report.maxTotalSeconds, stats.totalSeconds)
        }
        let sunday = cal.date(byAdding: .day, value: 6, to: monday) ?? monday
        report.startKey = Fmt.dateKey(monday)
        report.endKey = Fmt.dateKey(sunday)
        report.startLabel = Fmt.shortDateLabel(monday)
        report.endLabel = Fmt.shortDateLabel(sunday)
        return report
    }

    func monthlyStats(year: Int, month: Int) -> MonthReport {
        let cal = Fmt.calendar
        var components = DateComponents()
        components.year = year
        components.month = month + 1
        components.day = 1
        guard let first = cal.date(from: components),
              let range = cal.range(of: .day, in: .month, for: first) else { return MonthReport() }

        var report = MonthReport(
            key: Fmt.monthKey(year: year, month: month),
            label: Fmt.monthLabel(year: year, month: month),
            shortLabel: Fmt.monthShortLabel(year: year, month: month)
        )

        let leading = Fmt.mondayFirstIndex(first)
        for blank in 0..<leading {
            report.cells.append(MonthCell(id: blank, inMonth: false))
        }
        for dayNumber in range {
            guard let date = cal.date(byAdding: .day, value: dayNumber - 1, to: first) else { continue }
            let stats = statsForDay(Fmt.dateKey(date))
            report.cells.append(MonthCell(
                id: leading + dayNumber,
                inMonth: true,
                key: stats.key,
                dayNumber: dayNumber,
                isToday: stats.key == todayKey,
                focusSeconds: stats.focusSeconds,
                breakSeconds: stats.breakSeconds,
                sessions: stats.sessions
            ))
            report.sessions += stats.sessions
            report.focusSeconds += stats.focusSeconds
            report.breakSeconds += stats.breakSeconds
            report.maxDaySeconds = max(report.maxDaySeconds, stats.focusSeconds)
            if stats.focusSeconds > 0 { report.activeDays += 1 }
        }
        while report.cells.count % 7 != 0 {
            report.cells.append(MonthCell(id: 1000 + report.cells.count, inMonth: false))
        }
        return report
    }

    func allTimeStats() -> AllTimeReport {
        var report = AllTimeReport()
        var activeDays = Set<String>()
        var monthKeys = Set<String>()
        var streakDays = Set<String>()

        for entry in sessions {
            let active = entry.activeSeconds
            let day = Fmt.dateKey(entry.endedAt)
            if entry.phase == .focus {
                report.focusSeconds += active
                if active > 0 {
                    if !day.isEmpty {
                        activeDays.insert(day)
                        monthKeys.insert(String(day.prefix(7)))
                    }
                }
                if entry.isCompletedFocus {
                    report.sessions += 1
                    streakDays.insert(day)
                    if report.firstEndedAt == 0 || entry.endedAt < report.firstEndedAt {
                        report.firstEndedAt = entry.endedAt
                    }
                    if entry.endedAt > report.lastEndedAt { report.lastEndedAt = entry.endedAt }
                } else if active > 0 {
                    report.interrupted += 1
                }
            } else {
                report.breakSeconds += active
                if active > 0 {
                    report.breaks += 1
                    if !day.isEmpty { monthKeys.insert(String(day.prefix(7))) }
                }
                if entry.status != .completed && active > 0 { report.interrupted += 1 }
            }
        }
        report.activeDays = activeDays.count

        let ordered = streakDays.sorted()
        var currentRun = 0
        for (index, key) in ordered.enumerated() {
            if index > 0, let previous = Fmt.dayOrdinal(ordered[index - 1]),
               let current = Fmt.dayOrdinal(key), current == previous + 1 {
                currentRun += 1
            } else {
                currentRun = 1
            }
            report.longestStreak = max(report.longestStreak, currentRun)
        }

        let cal = Fmt.calendar
        var cursor = cal.startOfDay(for: currentDate)
        if !streakDays.contains(Fmt.dateKey(cursor)) {
            cursor = cal.date(byAdding: .day, value: -1, to: cursor) ?? cursor
        }
        while streakDays.contains(Fmt.dateKey(cursor)) {
            report.currentStreak += 1
            guard let previous = cal.date(byAdding: .day, value: -1, to: cursor) else { break }
            cursor = previous
        }

        // The chart shows at most the trailing twelve months of activity.
        let observed = monthKeys.sorted()
        if let firstKey = observed.first, let firstParts = Fmt.monthParts(firstKey) {
            let lastYear = cal.component(.year, from: currentDate)
            let lastMonth = cal.component(.month, from: currentDate) - 1
            let monthCount = (lastYear - firstParts.year) * 12 + lastMonth - firstParts.month + 1
            let firstIndex = max(0, monthCount - 12)
            if monthCount > 0 {
                for offset in firstIndex..<monthCount {
                    let absolute = firstParts.year * 12 + firstParts.month + offset
                    let monthReport = monthlyStats(year: absolute / 12, month: absolute % 12)
                    report.months.append(MonthBar(
                        key: monthReport.key,
                        label: monthReport.shortLabel,
                        focusSeconds: monthReport.focusSeconds,
                        breakSeconds: monthReport.breakSeconds,
                        sessions: monthReport.sessions
                    ))
                }
            }
        }
        report.maxChartSeconds = report.months.map(\.totalSeconds).max() ?? 0
        return report
    }

    /// Position of a timestamp within its day, 0…1, for the timeline.
    func timelineRatio(_ value: EpochMillis, key: String) -> Double {
        guard let start = Fmt.dayStartMillis(forKey: key) else { return 0 }
        return max(0, min(1, (value - start) / 86_400_000))
    }
}

File: windows/src/Pomodoro.Core/Reports.cs
namespace Pomodoro.Core;

public sealed record DayStats
{
    public string Key { get; init; } = string.Empty;
    public int Sessions { get; init; }
    public int FocusSeconds { get; init; }
    public int BreakSeconds { get; init; }
    public int Breaks { get; init; }
    public int CompletedBreaks { get; init; }
    public int ShortBreaks { get; init; }
    public int LongBreaks { get; init; }
    public int InterruptedFocus { get; init; }
    public int InterruptedBreaks { get; init; }
    public int Phases { get; init; }

    public int TotalSeconds => FocusSeconds + BreakSeconds;
    public string FocusText => Fmt.ReportDuration(FocusSeconds);
    public string BreakText => Fmt.ReportDuration(BreakSeconds);
    public string TotalText => Fmt.ReportDuration(TotalSeconds);
}

public sealed record WeekDay
{
    public string Key { get; init; } = string.Empty;
    public string Label { get; init; } = string.Empty;
    public int DayNumber { get; init; }
    public bool IsToday { get; init; }
    public int Sessions { get; init; }
    public int Breaks { get; init; }
    public int FocusSeconds { get; init; }
    public int BreakSeconds { get; init; }

    public int TotalSeconds => FocusSeconds + BreakSeconds;
    public string FocusText => Fmt.ReportDuration(FocusSeconds);
    public string BreakText => Fmt.ReportDuration(BreakSeconds);
}

public sealed record WeekReport
{
    public string StartKey { get; init; } = string.Empty;
    public string EndKey { get; init; } = string.Empty;
    public string StartLabel { get; init; } = string.Empty;
    public string EndLabel { get; init; } = string.Empty;
    public IReadOnlyList<WeekDay> Days { get; init; } = Array.Empty<WeekDay>();
    public int Sessions { get; init; }
    public int FocusSeconds { get; init; }
    public int BreakSeconds { get; init; }
    public int MaxTotalSeconds { get; init; }

    public string FocusText => Fmt.ReportDuration(FocusSeconds);
    public string BreakText => Fmt.ReportDuration(BreakSeconds);
    public string AverageDayText => Fmt.ReportDuration(FocusSeconds / 7.0);
}

public sealed record MonthCell
{
    public bool InMonth { get; init; }
    public string Key { get; init; } = string.Empty;
    public int DayNumber { get; init; }
    public bool IsToday { get; init; }
    public int FocusSeconds { get; init; }
    public int BreakSeconds { get; init; }
    public int Sessions { get; init; }

    public string FocusText => Fmt.ReportDuration(FocusSeconds);
}

public sealed record MonthReport
{
    public string Key { get; init; } = string.Empty;
    public string Label { get; init; } = string.Empty;
    public string ShortLabel { get; init; } = string.Empty;
    public IReadOnlyList<MonthCell> Cells { get; init; } = Array.Empty<MonthCell>();
    public int Sessions { get; init; }
    public int FocusSeconds { get; init; }
    public int BreakSeconds { get; init; }
    public int ActiveDays { get; init; }
    public int MaxDaySeconds { get; init; }

    public string FocusText => Fmt.ReportDuration(FocusSeconds);
    public string BreakText => Fmt.ReportDuration(BreakSeconds);
}

public sealed record MonthBar
{
    public string Key { get; init; } = string.Empty;
    public string Label { get; init; } = string.Empty;
    public int FocusSeconds { get; init; }
    public int BreakSeconds { get; init; }
    public int Sessions { get; init; }

    public int TotalSeconds => FocusSeconds + BreakSeconds;
}

public sealed record AllTimeReport
{
    public int Sessions { get; init; }
    public int FocusSeconds { get; init; }
    public int BreakSeconds { get; init; }
    public int Breaks { get; init; }
    public int ActiveDays { get; init; }
    public int CurrentStreak { get; init; }
    public int LongestStreak { get; init; }
    public int Interrupted { get; init; }
    public double FirstEndedAt { get; init; }
    public double LastEndedAt { get; init; }
    public IReadOnlyList<MonthBar> Months { get; init; } = Array.Empty<MonthBar>();
    public int MaxChartSeconds { get; init; }

    public string FocusText => Fmt.ReportDuration(FocusSeconds);
    public string BreakText => Fmt.ReportDuration(BreakSeconds);
    public string AverageSessionText =>
        Sessions > 0 ? Fmt.ReportDuration((double)FocusSeconds / Sessions) : "0m";
}

/// <summary>
/// Report aggregation, ported from the QML service. Every figure is derived
/// from <see cref="PomodoroService.Sessions"/>, so all front ends report
/// identical numbers from the same history file.
/// </summary>
public sealed partial class PomodoroService
{
    public DayStats StatsForDay(string key)
    {
        int sessions = 0, focus = 0, brk = 0, breaks = 0, completedBreaks = 0;
        int shortBreaks = 0, longBreaks = 0, interruptedFocus = 0, interruptedBreaks = 0, phases = 0;

        foreach (var entry in _sessions)
        {
            if (Fmt.DateKey(entry.EndedAt) != key) continue;
            var active = entry.ActiveSeconds;
            phases++;
            if (entry.Phase == Phase.Focus)
            {
                focus += active;
                if (entry.IsCompletedFocus) sessions++;
                else if (active > 0) interruptedFocus++;
            }
            else
            {
                brk += active;
                if (active > 0) breaks++;
                if (entry.Status == EntryStatus.Completed) completedBreaks++;
                if (entry.Phase == Phase.Short && active > 0) shortBreaks++;
                if (entry.Phase == Phase.Long && active > 0) longBreaks++;
                if (entry.Status != EntryStatus.Completed && active > 0) interruptedBreaks++;
            }
        }

        return new DayStats
        {
            Key = key,
            Sessions = sessions,
            FocusSeconds = focus,
            BreakSeconds = brk,
            Breaks = breaks,
            CompletedBreaks = completedBreaks,
            ShortBreaks = shortBreaks,
            LongBreaks = longBreaks,
            InterruptedFocus = interruptedFocus,
            InterruptedBreaks = interruptedBreaks,
            Phases = phases
        };
    }

    private static List<Segment> ClippedSegments(
        IReadOnlyList<Segment> segments, double rangeStart, double rangeEnd)
    {
        var result = new List<Segment>();
        foreach (var segment in segments)
        {
            var start = Math.Max(rangeStart, segment.StartedAt);
            var end = Math.Min(rangeEnd, segment.EndedAt);
            if (end <= start) continue;
            result.Add(new Segment(start, end));
        }
        return result;
    }

    /// <summary>
    /// Every phase overlapping the day, including the one in progress, sorted
    /// by start. This is what the timeline and the session list read.
    /// </summary>
    public List<SessionEntry> EntriesForDay(string key)
    {
        var dayStartMillis = Fmt.DayStartMillis(key);
        if (dayStartMillis is null) return [];
        var dayStart = dayStartMillis.Value;
        var dayEnd = dayStart + 86_400_000.0;

        var rows = new List<SessionEntry>();
        foreach (var entry in _sessions)
        {
            if (entry.EndedAt < dayStart || entry.StartedAt >= dayEnd) continue;
            rows.Add(entry with { Segments = ClippedSegments(entry.Segments, dayStart, dayEnd) });
        }

        if (PhaseStartedAt > 0)
        {
            var now = Fmt.NowMillis();
            var activeEnd = Math.Max(now, PhaseStartedAt);
            if (activeEnd >= dayStart && PhaseStartedAt < dayEnd)
            {
                rows.Add(new SessionEntry
                {
                    Id = "active-phase",
                    Phase = Phase,
                    Status = Running ? EntryStatus.Running : EntryStatus.Paused,
                    StartedAt = PhaseStartedAt,
                    EndedAt = now,
                    PlannedSeconds = _phasePlannedSeconds > 0 ? _phasePlannedSeconds : Duration(Phase),
                    ActiveSeconds = PhaseElapsed(now),
                    Segments = ClippedSegments(PhaseSegmentsAt(now), dayStart, dayEnd),
                    Note = Phase == Phase.Focus ? ActiveNote : string.Empty,
                    IsLive = true
                });
            }
        }

        // OrderBy is stable, matching JS Array.prototype.sort.
        return rows.OrderBy(row => row.StartedAt).ToList();
    }

    /// <summary>Completed focus sessions for the day, newest first.</summary>
    public List<SessionEntry> SessionsForDay(string key, int? limit = null)
    {
        var result = new List<SessionEntry>();
        var rows = EntriesForDay(key);
        for (var index = rows.Count - 1; index >= 0; index--)
        {
            if (!rows[index].IsCompletedFocus) continue;
            result.Add(rows[index]);
            if (limit is not null && result.Count >= limit.Value) break;
        }
        return result;
    }

    public WeekReport WeeklyStats(DateTime anchor)
    {
        var today = anchor.Date;
        var monday = today.AddDays(-Fmt.MondayFirstIndex(today));
        string[] names = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"];

        var days = new List<WeekDay>();
        int sessions = 0, focus = 0, brk = 0, maxTotal = 0;
        for (var index = 0; index < 7; index++)
        {
            var date = monday.AddDays(index);
            var stats = StatsForDay(Fmt.DateKey(date));
            days.Add(new WeekDay
            {
                Key = stats.Key,
                Label = names[index],
                DayNumber = date.Day,
                IsToday = stats.Key == TodayKey,
                Sessions = stats.Sessions,
                Breaks = stats.Breaks,
                FocusSeconds = stats.FocusSeconds,
                BreakSeconds = stats.BreakSeconds
            });
            sessions += stats.Sessions;
            focus += stats.FocusSeconds;
            brk += stats.BreakSeconds;
            maxTotal = Math.Max(maxTotal, stats.TotalSeconds);
        }

        var sunday = monday.AddDays(6);
        return new WeekReport
        {
            StartKey = Fmt.DateKey(monday),
            EndKey = Fmt.DateKey(sunday),
            StartLabel = Fmt.ShortDateLabel(monday),
            EndLabel = Fmt.ShortDateLabel(sunday),
            Days = days,
            Sessions = sessions,
            FocusSeconds = focus,
            BreakSeconds = brk,
            MaxTotalSeconds = maxTotal
        };
    }

    /// <param name="month">Zero-based, as the QML monthlyStats took it.</param>
    public MonthReport MonthlyStats(int year, int month)
    {
        var first = new DateTime(year, month + 1, 1, 0, 0, 0, DateTimeKind.Local);
        var daysInMonth = DateTime.DaysInMonth(year, month + 1);
        var leading = Fmt.MondayFirstIndex(first);

        var cells = new List<MonthCell>();
        for (var blank = 0; blank < leading; blank++) cells.Add(new MonthCell { InMonth = false });

        int sessions = 0, focus = 0, brk = 0, maxDay = 0, activeDays = 0;
        for (var dayNumber = 1; dayNumber <= daysInMonth; dayNumber++)
        {
            var date = first.AddDays(dayNumber - 1);
            var stats = StatsForDay(Fmt.DateKey(date));
            cells.Add(new MonthCell
            {
                InMonth = true,
                Key = stats.Key,
                DayNumber = dayNumber,
                IsToday = stats.Key == TodayKey,
                FocusSeconds = stats.FocusSeconds,
                BreakSeconds = stats.BreakSeconds,
                Sessions = stats.Sessions
            });
            sessions += stats.Sessions;
            focus += stats.FocusSeconds;
            brk += stats.BreakSeconds;
            maxDay = Math.Max(maxDay, stats.FocusSeconds);
            if (stats.FocusSeconds > 0) activeDays++;
        }
        while (cells.Count % 7 != 0) cells.Add(new MonthCell { InMonth = false });

        return new MonthReport
        {
            Key = Fmt.MonthKey(year, month),
            Label = Fmt.MonthLabel(year, month),
            ShortLabel = Fmt.MonthShortLabel(year, month),
            Cells = cells,
            Sessions = sessions,
            FocusSeconds = focus,
            BreakSeconds = brk,
            ActiveDays = activeDays,
            MaxDaySeconds = maxDay
        };
    }

    public AllTimeReport AllTimeStats()
    {
        var activeDays = new HashSet<string>();
        var monthKeys = new HashSet<string>();
        var streakDays = new HashSet<string>();
        int sessions = 0, focus = 0, brk = 0, breaks = 0, interrupted = 0;
        double firstEndedAt = 0, lastEndedAt = 0;

        foreach (var entry in _sessions)
        {
            var active = entry.ActiveSeconds;
            var day = Fmt.DateKey(entry.EndedAt);
            if (entry.Phase == Phase.Focus)
            {
                focus += active;
                if (active > 0 && day.Length > 0)
                {
                    activeDays.Add(day);
                    monthKeys.Add(day[..7]);
                }
                if (entry.IsCompletedFocus)
                {
                    sessions++;
                    streakDays.Add(day);
                    if (firstEndedAt == 0 || entry.EndedAt < firstEndedAt) firstEndedAt = entry.EndedAt;
                    if (entry.EndedAt > lastEndedAt) lastEndedAt = entry.EndedAt;
                }
                else if (active > 0)
                {
                    interrupted++;
                }
            }
            else
            {
                brk += active;
                if (active > 0)
                {
                    breaks++;
                    if (day.Length > 0) monthKeys.Add(day[..7]);
                }
                if (entry.Status != EntryStatus.Completed && active > 0) interrupted++;
            }
        }

        var ordered = streakDays.OrderBy(key => key, StringComparer.Ordinal).ToList();
        int longestStreak = 0, run = 0;
        for (var index = 0; index < ordered.Count; index++)
        {
            if (index > 0 &&
                Fmt.DayOrdinal(ordered[index - 1]) is int previous &&
                Fmt.DayOrdinal(ordered[index]) is int current &&
                current == previous + 1)
            {
                run++;
            }
            else
            {
                run = 1;
            }
            longestStreak = Math.Max(longestStreak, run);
        }

        var currentStreak = 0;
        var cursor = CurrentDate.Date;
        if (!streakDays.Contains(Fmt.DateKey(cursor))) cursor = cursor.AddDays(-1);
        while (streakDays.Contains(Fmt.DateKey(cursor)))
        {
            currentStreak++;
            cursor = cursor.AddDays(-1);
        }

        // The chart shows at most the trailing twelve months of activity.
        var months = new List<MonthBar>();
        var observed = monthKeys.OrderBy(key => key, StringComparer.Ordinal).ToList();
        if (observed.Count > 0 && Fmt.MonthParts(observed[0]) is { } firstParts)
        {
            var lastYear = CurrentDate.Year;
            var lastMonth = CurrentDate.Month - 1;
            var monthCount = (lastYear - firstParts.Year) * 12 + lastMonth - firstParts.Month + 1;
            var firstIndex = Math.Max(0, monthCount - 12);
            for (var offset = firstIndex; offset < monthCount; offset++)
            {
                var absolute = firstParts.Year * 12 + firstParts.Month + offset;
                var report = MonthlyStats(absolute / 12, absolute % 12);
                months.Add(new MonthBar
                {
                    Key = report.Key,
                    Label = report.ShortLabel,
                    FocusSeconds = report.FocusSeconds,
                    BreakSeconds = report.BreakSeconds,
                    Sessions = report.Sessions
                });
            }
        }

        return new AllTimeReport
        {
            Sessions = sessions,
            FocusSeconds = focus,
            BreakSeconds = brk,
            Breaks = breaks,
            ActiveDays = activeDays.Count,
            CurrentStreak = currentStreak,
            LongestStreak = longestStreak,
            Interrupted = interrupted,
            FirstEndedAt = firstEndedAt,
            LastEndedAt = lastEndedAt,
            Months = months,
            MaxChartSeconds = months.Count > 0 ? months.Max(month => month.TotalSeconds) : 0
        };
    }

    /// <summary>Position of a timestamp within its day, 0…1, for the timeline.</summary>
    public double TimelineRatio(double value, string key)
    {
        var start = Fmt.DayStartMillis(key);
        if (start is null) return 0;
        return Math.Max(0, Math.Min(1, (value - start.Value) / 86_400_000.0));
    }
}
